src/coreclr/src/gc/gc.cpp

```diff
@@ -31727,56 +31740,136 @@ void gc_heap::trim_youngest_desired_low_memory()
+
+    // we do a max of 160 kB per millisecond (160 MB per second) of pause time
+    // we don't want to take large pause times too seriously - limit to 10 seconds
+    ptrdiff_t max_decommit_size = min(ephemeral_elapsed, 10*1000) * 160 * 1024;
+    decommit_size = min(decommit_size, max_decommit_size);
+
+    slack_space = heap_segment_committed (ephemeral_heap_segment) - heap_segment_allocated (ephemeral_heap_segment) - decommit_size;
     decommit_heap_segment_pages (ephemeral_heap_segment, slack_space);
+#endif // !MULTIPLE_HEAPS
 
     gc_history_per_heap* current_gc_data_per_heap = get_gc_data_per_heap();
     current_gc_data_per_heap->extra_gen0_committed = heap_segment_committed (ephemeral_heap_segment) - heap_segment_allocated (ephemeral_heap_segment);
 }
 
+#ifdef MULTIPLE_HEAPS
+// return true if we actually decommitted anything
+bool gc_heap::decommit_step ()
+{
+    // should never get here for large pages because decommit_ephemeral_segment_pages
+    // will not do anything if use_large_pages_p is true
+    assert(!use_large_pages_p);
+
+    size_t decommit_size = 0;
+    for (int i = 0; i < n_heaps; i++)
+    {
+        gc_heap* hp = gc_heap::g_heaps[i];
+        decommit_size += hp->decommit_ephemeral_segment_pages_step ();
+    }
+    return (decommit_size != 0);
+}
+
+// return the decommitted size
+size_t gc_heap::decommit_ephemeral_segment_pages_step ()
+{
+    // we rely on desired allocation not being changed outside of GC
+    if (ephemeral_heap_segment->saved_desired_allocation != dd_desired_allocation(dynamic_data_of(0)))
+    {
+        GCToOSInterface::DebugBreak();
+    }
+    uint8_t* decommit_target = heap_segment_decommit_target (ephemeral_heap_segment);
+    size_t EXTRA_SPACE = 4 * OS_PAGE_SIZE;
+    decommit_target += EXTRA_SPACE;
+    uint8_t* committed = heap_segment_committed (ephemeral_heap_segment);
+    if (decommit_target < committed)
+    {
+        // we rely on other threads not messing with committed if we are about to trim it down
+        if (ephemeral_heap_segment->saved_committed != heap_segment_committed (ephemeral_heap_segment))
+        {
+            GCToOSInterface::DebugBreak();
+        }
+
+        const size_t DECOMMIT_STEP_SIZE = max (4096 / n_heaps, 100) * OS_PAGE_SIZE;
+        uint8_t* page_start = align_on_page ( max (decommit_target, committed - DECOMMIT_STEP_SIZE) );
+        size_t size = committed - page_start;
+        virtual_decommit(page_start, size, heap_number);
+        dprintf(3, ("Decommitting heap segment [%Ix, %Ix[(%d)",
+            (size_t)page_start,
+            (size_t)(page_start + size),
+            size));
+        heap_segment_committed (ephemeral_heap_segment) = page_start;
+        ephemeral_heap_segment->saved_committed = page_start;
+        if (heap_segment_used (ephemeral_heap_segment) > heap_segment_committed (ephemeral_heap_segment))
+        {
+            heap_segment_used (ephemeral_heap_segment) = heap_segment_committed (ephemeral_heap_segment);
```

Comment:
this is duplicating part of the code in `decommit_heap_segment_pages`; would be nice to have a worker function to do the common part so you don't need to care about details like updating used here. it's better to have that taken care of than having the person who's looking at this method later having to care about used.

Reply:
Done - common part is called decommit_heap_segment_pages_worker.